### DISCUSSION TITLE
Smart type check

### DISCUSSION THREAD

**Discussion by artelk** [dotnet/csharplang#5024] [Language Ideas]
I would like type checks to infer/modify the types like below
```cs
void GenericMethod<T>()
{
  IQueryable<T> q = ...;
  if(typeof(T) == typeof(Foo))
  {
    // inside this block the compiler could know that T==Foo and IQueryable<T>==IQueryable<Foo>
    // so you can avoid multiple casts
    q = q.Where(x => x.SomePropertyFromFoo == 42); // and you also can change the variable q assigning an instance of IQueryable<Foo>
  }

  //...
}

void OtherMethod()
{
  Base obj = ...;
  if(obj is Derived)
  {
    // inside this block the compiler could know that the type of the 'obj ' is Derived
    // so you can avoid using pattern matching in the 'if' condition introducing a new variable
    Console.WriteLine(obj.SomePropertyFromDerived);
  }

  //...
}
```

**Comment by HaloFour** [comment]
This would be a breaking change in that code that is interpreted in one way today will mean something completely different.

For example: [SharpLab](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ64GYAEicCCOA3ujuXrnACw4CyAFAJQ4C8AfHkgJwMBEBPkwDc6AL7pMVBDgBCOEIRJkK2HFACmAdzy1GLDl159ZQ0WgkY0cJADY8MmwHZlaCl3s0cAFQ0BnYAYiAEMWUjd3CgBLADMcBmCcKL85MJVIyOCAOn1zDPJLd0KKdI9demCoqGZXfJ9/QM0dWWYRUssxIA=)

**Reply by artelk** [reply]
That's true. But sometimes the language team introduce small breaking changes when they are useful. For example: https://davefancher.com/2012/11/03/c-5-0-breaking-changes/

**Reply by bernd5** [reply]
It would not be a small but huge change - and it would make method resultion very hard.

**Comment by ZacharyPatten** [comment]
https://github.com/dotnet/csharplang/discussions/3881

**Comment by FaustVX** [comment]
You could just use pattern matching :
```csharp
if (q is IQueryable<Foo> foo)
    q = foo. …
```
and
```csharp
if (obj is Derived d)
    d. …
```

**Reply by artelk** [reply]
You will also need to add cast to be able to assign q:
```cs
if (q is IQueryable<Foo> foo)
    q = (IQueryable<T>)foo. …
```
And it is probably impossible to use with anonymous types:
```cs
void GenericMethod<T>()
{
  IQueryable<T> tq = ...;
  var q = from t in tq
               join z in ...
               select new { t, z }; // some anonymous type that somehow depends on T
  if (q is IQueryable<???> foo)
  {
      q = (IQueryable<???>)foo. …
  }

  //...
}
```

**Reply by FaustVX** [reply]
Ok for your first point, but your second point doesn't works with your proposal

**Reply by artelk** [reply]
The second point (with Base/Derived classes) was added trying to make the things consistent. If the 'if' could infer the types making them more specific that should also work with 'is' expression.
With a new syntax for static type-check conditions the second point is not necessary.

There is a ConditionalAttribute for the methods. Maybe there could also be something like that which can be used inside methods with a new syntax (which also allows static type checks), so you would be able to write methods like this:
```cs
void GenericMethod<T>()
{
  IQueryable<T> tq = ...;
  var q = from t in tq
               join z in ...
               select new { t, z }; // some anonymous type that somehow depends on T
  static_conditional(T == Foo)
  {
     // here the compiler _really_ knows that T is Foo and doesn't require to put the ugly casts
      q = q.Where(o => o.t.SomePropertyFromFoo == o.z.Something); // e.g. it knows that o.t has property SomePropertyFromFoo and allows you to reassign the variable q like this
  }

  //...
}
```

**Comment by artelk** [comment]
To avoid breaking changes it can use a new syntax, something like this:
```cs
void GenericMethod<T>()
{
  IQueryable<T> tq = ...;
  var q = from t in tq
               join z in ...
               select new { t, z }; // some anonymous type that somehow depends on T
  //...

  if (typesAreTheSame(T, Foo))
  {
      // here all the variables that depend on T are implicitly casted
  }
}
```
Under the hood the compiler can safely use [Unsafe.As](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.unsafe.as?view=net-5.0#System_Runtime_CompilerServices_Unsafe_As__2___0__)  method so that will have even better performance than manual casts of every variable.

And manual explicit casts suppress static typing.
Imagine the code was like this:
```cs
if (q is IQueryable<Foo> foo)
    q = (IQueryable<T>) foo.Where(f => f.Prop1 == 42).Select(f => new Foo { ... });
```
Later someone changed that to:
```cs
if (q is IQueryable<Foo> foo)
    q = (IQueryable<T>) foo.Where(f => f.Prop1 == 42).Select(f => new Bar { ... }); // type changed
```
And the compiler is still happy but you have a type error now.


Samples of (awful) casts that could be avoided making the code more readable and type safe:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L70
https://github.com/dotnet/machinelearning/blob/main/src/Microsoft.ML.TensorFlow/TensorTypeExtensions.cs#L24
https://github.com/dotnet/ef6/blob/99c172bc465492fa37f9efa07900653975ec79b4/src/EntityFramework/ModelConfiguration/Conventions/Configuration/Property/KeyAttributeConvention.cs#L32
https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Dialogs/Prompts/NumberPrompt.cs#L137
